conf_file.cpp

```diff
@@ -16,268 +16,416 @@
+                              size_t table_len, void *data)
+{
+    struct section *s;
 
-    _trim(config->key);
-    _trim(config->value);
+    assert(section_name);
+    assert(table);
 
-    config->next = _current_section->configs;
-    _current_section->configs = config;
+    s = _find_section(section_name, strlen(section_name));
+    if (!s)
+        return -ENOENT;
 
-    return 0;
+    return _extract_options_from_section(s, table, table_len, data);
 }
 
-const char *ConfFile::next_from_section(const char *section_name, const char *key)
+struct config *ConfFile::_find_config(struct section *s, const char *key, size_t key_len)
 {
-    struct section *section;
-    struct config *config;
+    struct config *c;
 
-    if (!_current_section)
-        _current_section = _sections;
-
-    if (!_current_section) {
-        return NULL;
+    for (c = s->configs; c; c = c->next) {
+        if (c->key_len == key_len && strncasecmp(key, (char *)c->key, key_len) == 0)
+            return c;
     }
 
-    if (!strcaseeq(section_name, _current_section->name)) {
-        section = _sections;
+    return nullptr;
+}
 
-        while (section) {
-            if (strcaseeq(section->name, section_name)) {
-                _current_section = section;
-                _current_section->current_config = NULL;
-                break;
-            }
-            section = section->next;
-        }
+struct section *ConfFile::_find_section(const char *section_name, size_t len)
+{
+    struct section *s;
 
-        if (!section) {
-            return NULL;
-        }
+    for (s = _sections; s; s = s->next) {
+        if (memcaseeq(section_name, len, s->name, s->len))
+            return s;
     }
 
-    if (!_current_section->current_config
-        || !strcaseeq(_current_section->current_config->key, key)) {
-        config = _current_section->configs;
-    } else {
-        config = _current_section->current_config->next;
-    }
+    return nullptr;
+}
 
-    while (config) {
-        if (strcaseeq(config->key, key)) {
-            _current_section->current_config = config;
-            return config->value;
+int ConfFile::loop_sub_sections(const char *section_name, char separator, const char **sub_name,
+                                size_t *sub_len, const OptionsTable table[], size_t table_len,
+                                void *data, void **ptr)
+{
+    struct section *s;
+    size_t section_name_len;
+
+    assert(section_name);
+    assert(sub_name);
+    assert(sub_len);
+    assert(table);
+    assert(ptr);
+
+    section_name_len = strlen(section_name);
+
+    if (*ptr == NULL)
+        s = _sections;
+    else
+        s = ((section *)*ptr)->next;
+
+    for (; s; s = s->next) {
+        if (s->len - 1 > section_name_len && strncaseeq(section_name, s->name, section_name_len)
+            && s->name[section_name_len] == separator) {
+            *sub_name = s->name + section_name_len + 1;
+            *sub_len = s->len - (*sub_name - s->name);
```

Comment:
We don't allow unnamed sub sections because the if is checking if we have 2 chars (separator + 1) after the section name.
But, it is not trimming the subsection and section names, so we may have a subsection called " ".
So, do we need to trim section names, like:
[     section_name        sub_section_name       ]